### PR TITLE
Allow write/read unnamed tuple as nested Message in Protobuf format

### PR DESCRIPTION
**Pull request by Avogar**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow write/read unnamed tuple as nested Message in Protobuf format. Tuple elements and Message fields are mathced by position.

### CODE REVIEW

File: tests/queries/0_stateless/02707_protobuf_unnamed_tuple_as_nested_message.sh [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+message Nested {
+    int32 a = 1;
+    string b = 2;
+    repeated int32 c = 3;
+};
+
+message Message {
+    Nested x = 1;
+};' > 02707_schema_$CLICKHOUSE_TEST_UNIQUE_NAME.proto
```
Comment (vitlibar) [inline tests/queries/0_stateless/02707_protobuf_unnamed_tuple_as_nested_message.sh:18]: You can just put a ready *.proto file to `/tests/queries/0_stateless/format_schemas` with other test format schemas.

File: src/Formats/ProtobufSerializer.cpp [hunk 2]
```diff
@@ -3453,13 +3453,33 @@ namespace
...
+                            if (field_count != tuple_data_type.getElements().size())
+                                throw Exception(
+                                    ErrorCodes::NO_COLUMNS_SERIALIZED_TO_PROTOBUF_FIELDS,
+                                    "The number of fields in Protobuf message ({}) is not equal to the number of elements in unnamed Tuple ({})",
+                                    field_count,
+                                    tuple_data_type.getElements().size());
+                            for (size_t i = 0; i != field_count; ++i)
+                                element_names.push_back(field_descriptor.message_type()->field(static_cast<int>(i))->name());
+                        }
+
```
Comment (yakov-olkhovskiy) [inline src/Formats/ProtobufSerializer.cpp:3477]:
```suggestion
                    if (auto * message_type = field_descriptor.message_type())
                    {
                        Names element_names;
                        if (tuple_data_type.haveExplicitNames())
                        {
                            element_names = tuple_data_type.getElementNames();
                        }
                        else
                        {
                            /// Match unnamed Tuple elements and Message fields by position.
                            size_t field_count = message_type->field_count();
                            if (field_count != size_of_tuple)
                                throw Exception(
                                    ErrorCodes::NO_COLUMNS_SERIALIZED_TO_PROTOBUF_FIELDS,
                                    "The number of fields in Protobuf message ({}) is not equal to the number of elements in unnamed Tuple ({})",
                                    field_count,
                                    size_of_tuple);
                            for (size_t i = 0; i != field_count; ++i)
                                element_names.push_back(message_type->field(static_cast<int>(i))->name());
                        }

```

File: src/Formats/ProtobufSerializer.cpp [hunk 3]
```diff
@@ -3453,13 +3453,33 @@ namespace
                     const auto & tuple_data_type = assert_cast<const DataTypeTuple &>(*data_type);
                     size_t size_of_tuple = tuple_data_type.getElements().size();
 
-                    if (tuple_data_type.haveExplicitNames() && field_descriptor.message_type())
+                    if (field_descriptor.message_type())
```
Comment (yakov-olkhovskiy) [inline src/Formats/ProtobufSerializer.cpp:3456]:
```suggestion
                    if (auto * message_type = field_descriptor.message_type())
```
...and further all `field_descriptor.message_type()` can be replaced with `message_type`

File: src/Formats/ProtobufSerializer.cpp [hunk 4]
```diff
@@ -3453,13 +3453,33 @@ namespace
...
+                        else
+                        {
+                            /// Match unnamed Tuple elements and Message fields by position.
+                            size_t field_count = field_descriptor.message_type()->field_count();
+                            if (field_count != tuple_data_type.getElements().size())
+                                throw Exception(
+                                    ErrorCodes::NO_COLUMNS_SERIALIZED_TO_PROTOBUF_FIELDS,
+                                    "The number of fields in Protobuf message ({}) is not equal to the number of elements in unnamed Tuple ({})",
+                                    field_count,
+                                    tuple_data_type.getElements().size());
```
Comment (yakov-olkhovskiy) [inline src/Formats/ProtobufSerializer.cpp:3473]:
```suggestion
                            if (field_count != size_of_tuple)
                                throw Exception(
                                    ErrorCodes::NO_COLUMNS_SERIALIZED_TO_PROTOBUF_FIELDS,
                                    "The number of fields in Protobuf message ({}) is not equal to the number of elements in unnamed Tuple ({})",
                                    field_count,
                                    size_of_tuple);
```

**Review by yakov-olkhovskiy — APPROVED** [review]

**Pull request merged**